README
NAME
    OpusVL::AppKitX::SysParams -

DESCRIPTION
    UI Module for setting the SysParams.

METHODS
BUGS
AUTHOR
COPYRIGHT & LICENSE
    Copyright 2011 Opus Vision Limited, All Rights Reserved.

    This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

    If you require assistance, support, or further development of this software, please contact OpusVL using the details below:

    Telephone: +44 (0)1788 298 410
    Email: community@opusvl.com
    Web: http://opusvl.com



Changes
This file documents the revision history for Perl extension OpusVL::AppKit.

{{ $NEXT }}

0.29      2016-11-01 12:47:25+00:00 Europe/London

    - Added fontawesome files to this version, since AppKit doesn't have it by
      default

0.28      2016-06-09 12:02:43+01:00 Europe/London

    - Fixed saving of comments and labels

0.27      2016-06-09 11:16:58+01:00 Europe/London

    - Fixes to handling of data types, particularly when not set

0.26      2016-04-28 13:00:29+01:00 Europe/London

    - Fixed test db to get tests passing.

0.25      2016-03-14 12:55:59+00:00 Europe/London

    - Added tooltips to datatype selector.
    - Fixed bug preventing menu from working.

0.24      2016-03-12 10:47:30+00:00 Europe/London

    - Corrected version of dependency

0.23      2016-03-12 10:31:13+00:00 Europe/London

    - Fixed unique check so it should work with more versions of SQLite
    - Fixed packaging error.

0.22      2016-03-12 10:23:02+00:00 Europe/London

    - Converted to dist::zilla.
    - Support new OpusVL::SysParams with changes to data model.
    - New display and editing features with notional 'types'.

0.03    Added validation to prevent the inadvertent editing of parameters by 
        entering the same setting name when adding a setting.
0.02    Made UI use AppKit standards.  Added delete option and json set method.
0.01    Initial release.


t/lib/TestApp/Builder.pm
package TestApp::Builder;

use Moose;
extends 'OpusVL::AppKit::Builder';

override _build_superclasses => sub
{
  return [ 'OpusVL::AppKit' ]
};

override _build_plugins => sub {
    my $plugins = super(); # Get what CatalystX::AppBuilder gives you

    push @$plugins, qw(
        +OpusVL::AppKitX::SysParams
    );

    return $plugins;
};

override _build_config => sub {
    my $self   = shift;
    my $config = super(); # Get what CatalystX::AppBuilder gives you

    # point the AppKitAuth Model to the correct DB file....
    $config->{'Model::AppKitAuthDB'} = 
    {
        schema_class => 'OpusVL::AppKit::Schema::AppKitAuthDB',
        connect_info => [
          'dbi:SQLite:' . TestApp->path_to('root','appkit-auth.db'),
        ],
    };

    $config->{'Model::SysParams'} = 
    {
        connect_info => [
          'dbi:SQLite:' . TestApp->path_to('root','test_params.db'),
        ],
    };

    # .. add static dir into the config for Static::Simple..
    my $static_dirs = $config->{"Plugin::Static::Simple"}->{include_path};
    unshift(@$static_dirs, TestApp->path_to('root'));
    $config->{"Plugin::Static::Simple"}->{include_path} = $static_dirs;
    
    # .. allow access regardless of ACL rules...
    $config->{'appkit_can_access_actionpaths'} = ['custom/custom'];

    $config->{application_style} = 1;

    # DEBUGIN!!!!
    #$config->{'appkit_can_access_everything'} = 1;
    
    $config->{application_name} = 'AppKit TestApp';
    $config->{default_view}     = 'AppKitTT';
    
    return $config;
};

1;


Makefile.PL
use strict;
use warnings;
use inc::Module::Install 0.91;
use Module::Install::AuthorRequires;
use Module::Install::AuthorTests;

name 'OpusVL-AppKitX-SysParams';
all_from 'lib/OpusVL/AppKitX/SysParams.pm';

requires 'Moose';
requires 'namespace::autoclean';
requires 'HTML::FormFu::Model::DBIC';
requires 'OpusVL::AppKit';
requires 'OpusVL::SysParams';
requires 'JSON::MaybeXS';
requires 'List::UtilsBy';
requires 'Data::Munge';

build_requires 'Catalyst::Runtime' => '5.80015';
build_requires 'Test::WWW::Mechanize::Catalyst';
build_requires 'Test::More' => '0.88';

author_requires 'Test::Pod::Coverage' => '1.04';
author_requires 'Test::Pod' => '1.14';

author_tests 't/author';


if ($Module::Install::AUTHOR) {
    system("pod2text lib/OpusVL/AppKitX/SysParams.pm > README")
        and die $!;
}
resources (
    repository => {
        url => 'ssh://git/srv/git/OpusVL-AppKitX-SysParams',
        type => 'git',
    },
);


WriteAll();


lib/OpusVL/AppKitX/SysParams/Controller/SysInfo.pm
package OpusVL::AppKitX::SysParams::Controller::SysInfo;

use strict;
use Moose;
use namespace::autoclean;
use OpusVL::SysParams;
use Try::Tiny;
use JSON::MaybeXS;
use List::UtilsBy qw/zip_by/;

BEGIN { extends 'Catalyst::Controller::HTML::FormFu'; }
with 'OpusVL::AppKit::RolesFor::Controller::GUI';

__PACKAGE__->config
(
    appkit_name          => 'System Parameters',
    appkit_icon          => '/static/images/config-small.png',
    appkit_myclass       => 'OpusVL::AppKitX::SysParams',  
    appkit_shared_module => 'Configuration',
    appkit_method_group  => 'Configuration',
	path                 => 'adm/sysinfo',
);

sub auto 
    : Action
    : AppKitFeature('System Parameters')
{
	my $self = shift;
	my $c    = shift;

	$c->stash->{section}      = 'System Parameters';
    push @{$c->stash->{breadcrumbs}},
	{
        name => 'System Parameters',
        url  => $c->uri_for ($self->action_for ('list_params'))
    };

    push @{$c->stash->{header}->{css}}, '/static/modules/sysinfo/sysinfo.css';
    push @{$c->stash->{header}->{css}}, '/static/modules/sysinfo/fontawesome/font-awesome.min.css';
    push @{$c->stash->{header}->{js}}, '/static/modules/sysinfo/sysinfo.js';
    my $schema = $c->model('SysParams')->schema;
    $c->stash->{sys_params} = OpusVL::SysParams->new({ schema => $schema });

	$c->stash->{urls} =
	{
		sys_info_list => sub { $c->uri_for ( $self->action_for ('list_params')      ) },
		sys_info_set  => sub { $c->uri_for ( $self->action_for ('set_param'), shift ) },
		sys_info_set_ta  => sub { $c->uri_for ( $self->action_for ('set_textarea_param'), shift ) },
		sys_info_set_json  => sub { $c->uri_for ( $self->action_for ('set_json_param'), shift ) },
		sys_info_del  => sub { $c->uri_for ( $self->action_for ('del_param'), shift ) },
		sys_info_new  => sub { $c->uri_for ( $self->action_for ('new_param') ) },
        sys_info_comment => sub { $c->uri_for ($self->action_for ('set_comment'), shift ) },
	};
}

# FIXME: do we really want this to be Navigation Home?  I kind of suspect
# we either want to give this app a less generic name or allow it to be merged
# with other modules, in which case this navigation home could be a pain.
sub list_params
	: Path
	: NavigationName('System Parameters')
    : AppKitFeature('System Parameters')
{
	my $self = shift;
	my $c    = shift;
	
    my $grouped = $c->config->{'Model::SysParams'}->{group_sysparams};
	$c->stash->{sys_info} = $c->model('SysParams::SysInfo')->ordered;

    if ($grouped) {
        my $groups = {};
        for my $setting ($c->stash->{sys_info}->all) {
            my @path = split /\./, $setting->name;
            my $node = $groups;

            while (@path) {
                my $name = shift @path;
                my $path = $node->{path} || '';
                $node->{children}->{$name} //= {};

                $node = $node->{children}->{$name};
                $node->{path} ||= join '.', grep {$_} $path, $name;
            }

            $node->{param}   = $setting;
        }

        $c->stash->{sys_info} = $groups;
    
        $c->stash->{template} = 'modules/sysinfo/list_params_grouped.tt';
    }

    $c->stash->{pretty_json} = sub {
        JSON->new->pretty->ascii(0)->encode($_[0]);
    };
}

sub set_param
	: Path('set')
	: Args(1)
	: AppKitForm
    : AppKitFeature('System Parameters')
{
	my $self  = shift;
	my $c     = shift;
	my $name = shift;
    my $param = $c->model('SysParams::SysInfo')->find({
        name => $name
    });

	if ($c->req->param ('cancelbutton'))
	{
		$c->flash->{status_msg} = 'System Parameter not Changed';
		$c->res->redirect($c->stash->{urls}->{sys_info_list}->());
		$c->detach;
	}

    $self->_set_param($c, $param);

    $c->flash->{status_msg} = 'System Parameter Successfully Altered';
    $c->res->redirect($c->stash->{urls}{sys_info_list}->());
    $c->detach;
}


sub del_param
	: Path('del')
	: Args(1)
	: AppKitForm(delete.yml)
    : AppKitFeature('System Parameters')
{
	my $self  = shift;
	my $c     = shift;
	my $param = shift;
	my $form  = $c->stash->{form};
	my $value = $c->model ('SysParams::SysInfo')->get ($param);

	my $return_url = $c->stash->{urls}{sys_info_list}->();

	if ($c->req->param ('cancelbutton'))
	{
		$c->flash->{status_msg} = 'System Parameter Not Deleted';
		$c->res->redirect ($return_url);
		$c->detach;
	}

	if ($form->submitted_and_valid)
	{
		$c->model ('SysParams::SysInfo')->del ($param);
		$c->flash->{status_msg} = 'System Parameter Successfully Deleted';
		$c->res->redirect ($return_url);
		$c->detach;
	}
    else
    {
        $c->stash->{param_value} = $value;
        $c->stash->{param_name}  = $param;
    }
}

sub new_param
	: Path('new')
	: Args(0)
	: AppKitForm(modules/sysinfo/set_param.yml)
    : AppKitFeature('System Parameters')
{
	my $self  = shift;
	my $c     = shift;
	my $form  = $c->stash->{form};

	if ($c->req->param ('cancelbutton'))
	{
		$c->flash->{status_msg} = 'System Parameter Not Set';
		$c->res->redirect($c->stash->{urls}{sys_info_list}->());
		$c->detach;
	}
	
    $form->get_all_element('name')->type('Text');
    $form->process;
    $c->stash->{template} = 'modules/sysinfo/set_param.tt';

    $self->_set_param($c, $c->model('SysParams::SysInfo')->new_result({}));

    $c->flash->{status_msg} = 'System Parameter Successfully Created';
    $c->res->redirect($c->stash->{urls}{sys_info_list}->());
    $c->detach;
}

sub set_comment
    : Path('set_comment')
    : Args(1)
    : AppKitForm
    : AppKitFeature('System Parameters')
{
    my $self = shift;
    my $c = shift;
	my $name = shift;
	my $form  = $c->stash->{form};
    my $param = $c->model ('SysParams::SysInfo')->find_or_create({
        name => $name
    });

    my $return_url = $c->stash->{urls}->{sys_info_list}->();

	$form->default_values
	({
		name  => $name,
        comment => $param->comment
	});

    if ($form->submitted_and_valid) {
        $param->update({comment => $form->param_value('comment')});
		$c->flash->{status_msg} = "Comment updated";
		$c->res->redirect($return_url);
		$c->detach;
    }
}

sub _set_param {
    my ($self, $c, $param) = @_;

    my $form = $c->stash->{form};

    # this for ordering purposes.
    my $data_types = $c->model('SysParams::SysInfo')
        ->result_source
        ->column_info('data_type')
        ->{extra}
        ->{list};

    my %data_type_options = zip_by { @_ }
        $data_types,
        $c->model('SysParams::SysInfo')->result_source->column_info('data_type')->{extra}->{labels}
    ;
    my %actual_options = map { $_ => $data_type_options{$_} } @{ $param->viable_type_conversions };

    $form->get_all_element({ name => 'data_type' })->options([
        map { +{
            value => $_,
            label => $actual_options{$_},
            label_attributes => { title => $actual_options{$_} }
        } } 
        grep { exists $actual_options{$_} } @$data_types
    ]);
    $form->process;

	my $return_url = $c->stash->{urls}{sys_info_list}->();

	if ($form->submitted_and_valid)
	{
        my $type = $c->req->body_params->{'data_type'};
        my $name = $c->req->body_params->{'name'};

        my $update = {
            name => $name,
            data_type => $type,
            label => $c->req->body_params->{'label'},
            comment => $c->req->body_params->{'comment'},
        };
        if ($type eq 'object') {
            $update->{value} = $c->req->body_params->{value_json};
        }
        elsif ($type eq 'boolean') {
            my $tf = $c->req->body_params->{value} ? JSON->true : JSON->false;
            $update->{value} = JSON->new->allow_nonref->encode($tf);
        }
        else {
            $update->{value} = JSON->new->allow_nonref->encode($c->req->body_params->{value});
        }
        $param->set_columns($update);

        my $updated_ok = try {
            if ($param->in_storage) {
                $param->update;
            }
            else {
                $param->insert;
            }
            1;
        }
        catch {
            $c->log->debug(__PACKAGE__ . '->set_json_param exception: ' . $_);

            if (/UNIQUE/i) {
                $c->stash->{error_msg} = "Parameter $name already exists";
            }
            else {
                $form->get_field('value')->get_constraint({ type => 'Callback' })->force_errors(1);
            }
            $form->process;
            0;
        };

        return if $updated_ok;
	}

    # Make sure a suitable type is available, for old entries
    $param->set_type_from_value
        if ! $param->data_type
        and defined $param->value;

	$form->default_values
	({
		name  => $param->name,
        value => scalar $param->decoded_value,
        label => $param->label,
        comment => $param->comment,
        data_type => $param->data_type // 'text',
	});

    $c->stash->{param} = $param;
    $c->stash->{pretty_json} = sub {
        JSON->new->allow_nonref->pretty->ascii(0)->encode($_[0]);
    };
    $c->detach;
}

1;


=head1 NAME

OpusVL::AppKitX::SysParams::Controller::SysInfo

=head1 DESCRIPTION

=head1 METHODS

=head2 auto

=head2 list_params

=head2 set_textarea_param

=head2 set_param

=head2 set_json_param

=head2 del_param

=head2 new_param

=head2 set_comment

=head1 ATTRIBUTES


=head1 LICENSE AND COPYRIGHT

Copyright 2012 OpusVL.

This software is licensed according to the "IP Assignment Schedule" provided with the development project.

=cut


lib/auto/OpusVL/AppKitX/SysParams/root/static/modules/sysinfo/sysinfo.js
$(function() {
    $('.field.field-type').on('change', setFieldType('.field.field-value'));
    $(document).on('input', '.field.field-value', setTypeOptions('.field.field-type'));
    $(document).on('input', '.field.field-value', setAllFields);

    // I can't work with FormFu so I'm doing this.
    // We won't need this in the FB11 version.
    var input = {};
    input.textarea = $('.templates .js-value-textarea').detach();
    input.text = $('.templates .js-value-text').detach();
    input.object = $('.templates .js-value-raw').detach();
    input.boolean = $('.templates .js-value-boolean').detach();
    input.array = $('.templates .js-value-array').detach();

    // Pretend to fire an event
    setFieldType('.field.field-value')({ srcElement: $('.field.field-type :checked')[0] });
    setTypeOptions('.field.field-type')({ srcElement: $('.field.field-value')[0] });

    $(document).on('input', '.field-value-new', function() {
        var $this = $(this);

        $this.prev('.remove').clone().appendTo($this.parent());
        $this.clone().val('').appendTo($this.parent());

        $this
            .removeClass('field-value-new')
            .addClass('field-value-' + ($this.parent().find('input').length - 1))
            .attr('name', 'value');
    });

    $(document).on('click', '.remove', function() { $(this).prev().remove(); $(this).remove() });

    function setFieldType(selector) {
        if (! $(selector).length) {
            return function() {};
        }

        return function(event) {
            var $field = $(selector);
            var type = event.srcElement.value;

            var $newField = input[type];
            var c = $field.attr('class');
            $field.removeAttr('class');
            $newField.attr('class', c);

            $field.replaceWith($newField);
        }
    }

    function setTypeOptions(selector) {
        if (! $(selector).length) {
            return function() {};
        }
        return function(event) {
            var $currentTypeField = $(selector).find(':checked');
            var $field = $(event.srcElement);

            // if the JSON field is an object or array it cannot be converted
            if ($currentTypeField.val() == 'json') {
                if ($field.val().match(/^(\{|\[)/)) {
                    $('.field.field-type [value=multi], .field.field-type [value=string').attr('disabled', true);
                }
                else {
                    $('.field.field-type [value=multi], .field.field-type [value=string').attr('disabled', false);
                }
            }
        }
    }

    function setAllFields(event) {
        var $field = $(event.srcElement);
        var val = $field.val();
        var $currentTypeField = $('.field.field-type :checked');


        if ($currentTypeField.val() == 'json') {
            // mid-edit this might fail.
            try {
                val = JSON.parse(val);
                input.multi.val(val);
                input.string.val(val);
            } catch (e) {}
        }
        else {
            input.multi.val(val);
            input.string.val(val);
            input.json.val(JSON.stringify(val));
        }
    }
});


lib/OpusVL/AppKitX/SysParams.pm
package OpusVL::AppKitX::SysParams;
use Moose::Role;
use CatalystX::InjectComponent;
use File::ShareDir qw/module_dir/;
use namespace::autoclean;
use experimental 'smartmatch';

our $VERSION = '0.30';

after 'setup_components' => sub {
    my $class = shift;
   
    my $tt_view       = $class->config->{default_view} || 'TT';
    my $template_path = module_dir('OpusVL::AppKitX::SysParams') . 
                        '/root/templates';
    unless ($class->view($tt_view)->include_path ~~ $template_path) {
        push @{$class->view($tt_view)->include_path}, $template_path;
    }
   
    # .. add static dir into the config for Static::Simple..
    my $static_dirs = $class->config->{"Plugin::Static::Simple"}->{include_path};
    unshift(@$static_dirs, File::Spec->rel2abs(module_dir(__PACKAGE__) . '/root' ));
    $class->config->{"Plugin::Static::Simple"}->{include_path} = $static_dirs;
    
    # .. inject your components here ..
    CatalystX::InjectComponent->inject(
        into      => $class,
        component => 'OpusVL::AppKitX::SysParams::Controller::SysInfo',
        as        => 'Controller::Modules::SysInfo'
    );

    CatalystX::InjectComponent->inject(
        into      => $class,
        component => 'OpusVL::AppKitX::SysParams::Model::SysParams',
        as        => 'Model::SysParams'
    );
};

1;

=head1 NAME

OpusVL::AppKitX::SysParams - UI for SysParams.

=head1 DESCRIPTION

UI Module for setting the SysParams.

=head1 METHODS

=head1 BUGS

=head1 AUTHOR

=head1 COPYRIGHT & LICENSE

Copyright 2011 Opus Vision Limited.

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

If you require assistance, support, or further development of this software, please contact OpusVL using the details below:

=over 4

=item *

Telephone: +44 (0)1788 298 410

=item *

Email: community@opusvl.com

=item *

Web: L<http://opusvl.com>

=back

=cut

